ocaml/xenopsd/xc/mem_stats.ml

```diff
@@ -0,0 +1,391 @@
+     * writing a VT_Unknown causes an exception and prevents other values from being written too
+     * the way to write an unknown value is to write something out of min/max range, which then gets
+     converted to nan upon reading (or write nan if the type is float, don't do this if type is int
+     since it causes a metadata change)
+     * use Sys.word_size, and not hard-coded word-to-kib conversion, the old code still had the
+     values for a 32-bit word
+     * the value type (int64 vs float) is cached by the transport, changing it causes a metadata
+     change
+     * it is best to always write out the same Rrd list, even if some values are missing, otherwise
+     the metadata would have to get reparsed (crc changed). But that doesn't actually work that way
+     because the metadata also includes the values (they are changed to uninitialized_ds only upon
+     read), thus any change in value triggers a metadata crc change, and a metadata reparse! Still
+     lets assume this will be fixed and avoid needless metadata changes!
+     * everything gets converted to a float in the end: int64 type just has some more precision prior
+     to delta calculation. However for our purposes here the 'float' type (which is double precision)
+     has 53-bits of precision, which is more than enough for KiB - it results in 63-bits of precision
+     in term of bytes (including sign)
+  *)
+
+  let ds_int64 i = Rrd.VT_Int64 i
+
+  let ds_float f = Rrd.VT_Float f
+
+  let ds_update ds v = {ds with Ds.ds_value= v}
+
+  let executable = Filename.basename Sys.executable_name
+
+  let ds_name name =
+    Printf.sprintf "%s_%s" executable (Astring.String.Ascii.lowercase name)
+
+  let ds_description desc = Printf.sprintf desc executable
+
+  let to_float ds =
+    match ds.Ds.ds_value with
+    | Rrd.VT_Int64 i ->
+        if i >= 0L then Int64.to_float i else nan
+    | Rrd.VT_Float f ->
+        f
+    | Rrd.VT_Unknown ->
+        nan
+
+  let define_unit ~ty ~units ~min ?(default = true) name description =
+    let name = ds_name name in
+    let description = ds_description description in
+    fun value ->
+      Ds.ds_make ~name ~description ~value ~ty ~default ~units ~min ()
+
+  let kib ?default ?(min = 0.0) v =
+    define_unit ~ty:Rrd.Gauge ~units:"KiB" ~min ?default v
+
+  let kib_per_s = define_unit ~ty:Rrd.Derive ~units:"KiB/s" ~min:0.0
+
+  let count = define_unit ~ty:Rrd.Gauge ~units:"" ~min:0.0
+
+  let words_to_kib w = w *. float (Sys.word_size / 8) /. 1024. |> ds_float
+end
+
+module Proc = struct
+  module KV = Astring.String.Map
+
+  (** [drop_unit s] parses values of the form 'NN kB' or 'NN' and returns 'NN'. *)
+  let drop_unit s =
+    match Astring.String.cut ~sep:" " s with None -> s | Some (v, _) -> v
+
+  (** [parse_value s] parses values of the form '  N kB' or '    N'. *)
+  let parse_value s = s |> Astring.String.trim |> drop_unit |> Int64.of_string
```

Comment:
I would use `sscanf`

Reply:
Although I could potentially have 2 parsing functions here, one for the lines ending in kB, another for the ones not.
I'll do that as a separate commit on top, so that we can go back to this one in case anything breaks with the new functions.

Reply:
https://github.com/xapi-project/xen-api/blob/a836eb3f3c2b63f62978cb1b5ee4906185f3e590/ocaml/xapi-cli-server/record_util.ml#L1075-L1083 The unit can be optional if you check the matched string

Reply:
Updated